cloud/pkg/cloudhub/handler/messagehandler.go

```diff
@@ -142,6 +142,9 @@ func (mh *MessageHandle) OnRegister(connection conn.Connection) {
 
 	if _, ok := mh.KeepaliveChannel[nodeID]; !ok {
 		mh.KeepaliveChannel[nodeID] = make(chan struct{}, 1)
+	} else {
+		klog.Warningf("Node %v/%v has not yet exit, OnRegister failed", projectID, nodeID)
```

Comment:
has not yet exit?

Reply:
How about this "Node %v/%v is already registered, OnRegister again failed" ?

Reply:
![Uploading 20200518062209.png…]()
According to #1559 , this bug is due to bad network. Call the function OnRegister() in a short time make the channel 2 times by func (mh *MessageHandle) OnRegister(connection conn.Connection) and with same nodeID. Ideally, the life cycle of a channel looks like this: 2 make operation produces  corresponding connection with different nodeID in KeepaliveChannel,  and close the connection without disturbing each other. The key value in map KeepaliveChannel[nodeID] is nodeID. Therefore, at this time, the two make operations have only one channel and two close operations, which will cause the error of closing the nil channel. This make operation should be atomic. 
Therefore, when registering a node use func OnRegister(connection conn.Connection),  we should make sure to exit the previous channel with the same nodeID, otherwise it will not be registered.

Reply:
@fisherxu

Reply:
Based on further investigation #1704 #1670, the problem may not be limited to atomization of registerNode and unRegisterNode